File: src/main/java/dev/sky_lock/mocar/listener/ChunkEventListener.java
package dev.sky_lock.mocar.listener;

import dev.sky_lock.mocar.car.Car;
import dev.sky_lock.mocar.car.CarArmorStand;
import dev.sky_lock.mocar.car.CarEntities;
import org.bukkit.Chunk;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.world.ChunkLoadEvent;
import org.bukkit.event.world.ChunkUnloadEvent;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author sky_lock
 */

public class ChunkEventListener implements Listener {
    private final List<Car> cars = new ArrayList<>();

    @EventHandler
    public void onChunkUnload(ChunkUnloadEvent event) {
        Arrays.stream(event.getChunk().getEntities())
                .filter(entity -> entity instanceof CarArmorStand.CraftCar)
                .forEach(entity -> {
                    Car car = CarEntities.getCar(((CarArmorStand) ((CarArmorStand.CraftCar) entity).getHandle()));
                    car.kill();
                    cars.add(car);
                });
    }

    @EventHandler
    public void onChunkLoad(ChunkLoadEvent event) {
        cars.stream().filter(car -> isSameChunk(car.getLocation().getChunk(), event.getChunk()))
                .collect(Collectors.toList()).removeIf(CarEntities::spawn);
    }

    private boolean isSameChunk(Chunk chunk1, Chunk chunk2) {
        return chunk1.getX() == chunk2.getX() && chunk1.getZ() == chunk2.getZ();
    }

}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/listener/ChunkEventListener.kt
package dev.sky_lock.pocketlifevehicle.listener

import dev.sky_lock.pocketlifevehicle.vehicle.entity.EntityVehicleFacade
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.world.ChunkLoadEvent
import org.bukkit.event.world.ChunkUnloadEvent

/**
 * @author sky_lock
 */
class ChunkEventListener : Listener {

    @EventHandler
    fun onChunkUnload(event: ChunkUnloadEvent) {
        event.chunk.entities
                .forEach { entity ->
                    val vehicle = EntityVehicleFacade.fromBukkit(entity) ?: return@forEach
                    vehicle.unload()
                }
    }

    @EventHandler
    fun onChunkLoad(event: ChunkLoadEvent) {
        event.chunk.entities
            .forEach { entity ->
                val vehicle = EntityVehicleFacade.fromBukkit(entity) ?: return@forEach
                vehicle.load()
            }
    }
}